### PR TITLE
deprecate GIL refs in function argument

### PR DESCRIPTION
**Pull request by davidhewitt**
This was built on-stream today, the goal is to add a deprecation warning to `#[pyfunction]` and `#[pymethods]` arguments which take a GIL ref. The span is wrong but otherwise this technique works. The code is also a bit of a bodge job as I was floundering a bit on-stream trying to figure out how to get it to work 😂 

This is _not_ in a reviewable state but I felt like pushing the code anyway so it can be found easily when I get a chance to tidy it up.

### CODE REVIEW

**Comment by LilyFoote** [comment]
Will we get warnings from `#[pymodule]` with this?

**Comment by davidhewitt** [comment]
No, and also not for `pass_module` or `slf: &PyCell` I think. I will add all those to the tracking issue.

**Comment by davidhewitt** [comment]
Ok, with luck, this now passes CI!

File: src/instance.rs [hunk 1]
```diff
@@ -511,31 +511,8 @@ unsafe impl<T> AsPyPointer for Bound<'_, T> {
...
-    ///     let bound = borrowed.to_owned();
-    ///     drop(tuple);
-    ///
-    ///     assert_eq!(bound.extract::<i32>().unwrap(), 1);
-    ///     Ok(())
-    /// })
-    /// # }
-    pub fn to_owned(self) -> Bound<'py, T> {
+    /// Creates a new owned `Bound` from this borrowed reference by increasing the reference count.
+    pub(crate) fn to_owned(self) -> Bound<'py, T> {
```
Comment (Dr-Emann) [inline src/instance.rs:515]: Think this accidentally reverts https://github.com/PyO3/pyo3/pull/3963/files

Comment (davidhewitt) [reply src/instance.rs:515]: Thanks, good catch!

File: tests/test_proto_methods.rs [hunk 2]
```diff
@@ -767,18 +767,18 @@ impl DescrCounter {
     /// Each access will increase the count
     fn __get__<'a>(
         mut slf: PyRefMut<'a, Self>,
-        _instance: &PyAny,
+        _instance: &Bound<'_, PyAny>,
         _owner: Option<&PyType>,
```
Comment (Icxolu) [inline tests/test_proto_methods.rs:771]:
```suggestion
        _owner: Option<&Bound<'_, PyType>>,
```

File: pyo3-macros-backend/src/params.rs [hunk 3]
```diff
@@ -9,6 +9,56 @@ use quote::{quote, quote_spanned};
...
+    }
+
+    pub fn check_gil_refs(&self) -> TokenStream {
+        let gil_refs_checkers = self
+            .gil_refs_checkers
+            .iter()
+            .map(|e| quote_spanned! { e.span() => #e.function_arg(); });
+        quote! {
+            #(#gil_refs_checkers)*
+        }
```
Comment (Icxolu) [inline pyo3-macros-backend/src/params.rs:58]:
```suggestion
       self.gil_refs_checkers
            .iter()
            .map(|e| quote_spanned! { e.span() => #e.function_arg(); })
            .collect()
```

I think you can also just `.collect()` into a `TokenStream`.

File: tests/ui/invalid_frozen_pyclass_borrow.stderr [hunk 4]
```diff
@@ -16,6 +16,19 @@ note: required by a bound in `extract_pyclass_ref_mut`
...
+9 | #[pymethods]
+  | ^^^^^^^^^^^^ expected `False`, found `True`
+  |
+note: required by a bound in `PyRefMut`
+ --> src/pycell.rs
+  |
+  | pub struct PyRefMut<'p, T: PyClass<Frozen = False>> {
+  |                                    ^^^^^^^^^^^^^^ required by this bound in `PyRefMut`
+  = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)
+
```
Comment (Icxolu) [inline tests/ui/invalid_frozen_pyclass_borrow.stderr:31]: Given the corresponding test did not change i'm not sure where this one comes from and if it is expected

File: tests/ui/invalid_pymethod_enum.stderr [hunk 5]
```diff
@@ -9,3 +9,16 @@ note: required by a bound in `extract_pyclass_ref_mut`
...
+  |
+9 | #[pymethods]
+  | ^^^^^^^^^^^^ expected `False`, found `True`
+  |
+note: required by a bound in `PyRefMut`
+ --> src/pycell.rs
+  |
+  | pub struct PyRefMut<'p, T: PyClass<Frozen = False>> {
+  |                                    ^^^^^^^^^^^^^^ required by this bound in `PyRefMut`
+  = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Comment (Icxolu) [inline tests/ui/invalid_pymethod_enum.stderr:24]: Same here, is this expected?

**Review by Icxolu — COMMENTED** [review]
Very nice 🚀! I really like how `Holders` turned out, and how it could be reused to handle the `self_arg` as well. I found a few places to double check on (plus some very minor nits).

File: pytests/src/pyclasses.rs [hunk 6]
```diff
@@ -65,7 +65,7 @@ struct AssertingBaseClassGilRef;
 impl AssertingBaseClassGilRef {
     #[new]
     #[classmethod]
-    fn new(cls: &PyType, expected_type: &PyType) -> PyResult<Self> {
+    fn new(cls: &PyType, expected_type: &Bound<'_, PyType>) -> PyResult<Self> {
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply pytests/src/pyclasses.rs:68]: Ah indeed I just was mechanically fixing all the warnings, good catch 👀

File: tests/ui/invalid_frozen_pyclass_borrow.stderr (hunk 4 above)
Comment (davidhewitt) [reply tests/ui/invalid_frozen_pyclass_borrow.stderr:31]: I think what's happened is that the spans for the code in question are now mixed, so the error has duplicated onto both spans.

It's not really expected but I think it'll also be quite hard to fix without a bit of trial-and-error on spans. I think for now I'm inclined to leave this as-is and hope when we remove these warnings in a release or two it cleans up again. It is a touch ugly though. :(

File: tests/ui/invalid_pymethod_enum.stderr (hunk 5 above)
Comment (davidhewitt) [reply tests/ui/invalid_pymethod_enum.stderr:24]: Same as above :(

File: tests/ui/invalid_frozen_pyclass_borrow.stderr (hunk 4 above)
Comment (Icxolu) [reply tests/ui/invalid_frozen_pyclass_borrow.stderr:31]: I see, sounds reasonable

**Review by Icxolu — APPROVED** [review]

**Pull request merged**